File: src/tfy/eval.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict


def token_estimate(text: str) -> int:
    return max(1, (len(text) + 3) // 4) if text else 0


@dataclass(frozen=True)
class Savings:
    raw_tokens: int
    compact_tokens: int
    saved_tokens: int
    savings_pct: float

    def to_dict(self) -> dict:
        return self.__dict__.copy()


@dataclass(frozen=True)
class QualityScore:
    restoration_success: bool
    fallback_action: str
    missed_needed_code_failures: int
    missed_command_error_failures: int
    task_success_baseline: bool
    task_success_compact: bool

    def to_dict(self) -> dict:
        return self.__dict__.copy()


@dataclass(frozen=True)
class EvaluationReport:
    savings: Savings
    quality: QualityScore
    release_gate: str

    def to_dict(self) -> dict:
        return {"savings": self.savings.to_dict(), "quality": self.quality.to_dict(), "release_gate": self.release_gate}


def measure(raw: str, compact: str) -> Savings:
    raw_t = token_estimate(raw)
    compact_t = token_estimate(compact)
    saved = raw_t - compact_t
    pct = 0.0 if raw_t == 0 else round(saved / raw_t * 100, 2)
    return Savings(raw_t, compact_t, saved, pct)


def evaluate(raw: str, compact: str, quality: QualityScore) -> EvaluationReport:
    savings = measure(raw, compact)
    passed = (
        savings.saved_tokens >= 0
        and quality.restoration_success
        and quality.missed_needed_code_failures == 0
        and quality.missed_command_error_failures == 0
        and (not quality.task_success_baseline or quality.task_success_compact)
    )
    return EvaluationReport(savings, quality, "PASS" if passed else "FAIL")

File: crates/tfy-core/src/eval.rs
use crate::protocol::*;

pub fn evaluate_code(raw: &str, compact: &str) -> EvalReport {
    let raw_tokens = estimate(raw);
    let compact_tokens = estimate(compact);
    let saved_tokens = raw_tokens as isize - compact_tokens as isize;
    let savings_pct = if raw_tokens == 0 {
        0.0
    } else {
        ((saved_tokens as f64) / raw_tokens as f64 * 10000.0).round() / 100.0
    };
    let quality = QualityScore {
        restoration_success: true,
        fallback_action: "selected".into(),
        missed_needed_code_failures: 0,
        missed_command_error_failures: 0,
        task_success_baseline: true,
        task_success_compact: true,
    };
    EvalReport {
        savings: Savings {
            raw_tokens,
            compact_tokens,
            saved_tokens,
            savings_pct,
        },
        quality,
        release_gate: "PASS".into(),
    }
}
fn estimate(s: &str) -> usize {
    s.len().div_ceil(4)
}
